dataplane/saiserver/l2.go

```diff
@@ -39,25 +50,147 @@ func newL2mcGroup(mgr *attrmgr.AttrMgr, dataplane switchDataplaneAPI, s *grpc.Se
+		return nil, err
+	}
+	// Update internal data
+	if mg.groups[req.GetL2McGroupId()] == nil {
+		return nil, fmt.Errorf("L2MC group id %q not found", req.GetL2McGroupId())
+	}
+	if _, ok := mg.groups[req.GetL2McGroupId()][req.GetL2McOutputId()]; ok {
+		return nil, fmt.Errorf("found existing member %q", req.GetL2McOutputId())
+	}
+	mg.groups[req.GetL2McGroupId()][req.GetL2McOutputId()] = &l2mcGroupMember{
+		oid:      id,
+		outputId: req.GetL2McOutputId(),
+		groupId:  req.GetL2McGroupId(),
+	}
+	// Update L2MC Group member attributes
+	attr := &saipb.L2McGroupMemberAttribute{
+		L2McGroupId:  req.L2McGroupId,
+		L2McOutputId: req.L2McOutputId,
+	}
+	mg.mgr.StoreAttributes(id, attr)
+	// Update L2MC Group Attributes.
+	gReq := &saipb.GetL2McGroupAttributeRequest{Oid: req.GetL2McGroupId(), AttrType: []saipb.L2McGroupAttr{saipb.L2McGroupAttr_L2MC_GROUP_ATTR_L2MC_MEMBER_LIST, saipb.L2McGroupAttr_L2MC_GROUP_ATTR_L2MC_OUTPUT_COUNT}}
+	gResp := &saipb.GetL2McGroupAttributeResponse{}
+	if err := mg.mgr.PopulateAttributes(gReq, gResp); err != nil {
+		return nil, err
+	}
+	gAttrs := gResp.GetAttr()
+	gAttrs.L2McMemberList = append(gAttrs.GetL2McMemberList(), id)
+	*gAttrs.L2McOutputCount += 1
+	mg.mgr.StoreAttributes(req.GetL2McGroupId(), gAttrs)
+	// Update Switch Attributes.
+	swReq := &saipb.GetSwitchAttributeRequest{Oid: req.GetSwitch(), AttrType: []saipb.SwitchAttr{saipb.SwitchAttr_SWITCH_ATTR_AVAILABLE_L2MC_ENTRY}}
+	swResp := &saipb.GetSwitchAttributeResponse{}
+	if err := mg.mgr.PopulateAttributes(swReq, swResp); err != nil {
+		return nil, err
+	}
+	attrs := swResp.GetAttr()
+	*attrs.AvailableL2McEntry = attrs.GetAvailableL2McEntry() - 1
+	mg.mgr.StoreAttributes(req.GetSwitch(), attrs)
 	return &saipb.CreateL2McGroupMemberResponse{Oid: id}, nil
 }
 
-// TODO: Implement this.
-func (mg *l2mcGroup) RemoveL2McGroup(context.Context, *saipb.RemoveL2McGroupRequest) (*saipb.RemoveL2McGroupResponse, error) {
+func (mg *l2mcGroup) RemoveL2McGroup(ctx context.Context, req *saipb.RemoveL2McGroupRequest) (*saipb.RemoveL2McGroupResponse, error) {
+	if mg.groups[req.GetOid()] == nil {
+		return nil, fmt.Errorf("cannot find group %q", req.GetOid())
+	}
+	// Remove all members in the group
+	for _, p := range mg.groups[req.GetOid()] {
+		_, err := attrmgr.InvokeAndSave(ctx, mg.mgr, mg.RemoveL2McGroupMember, &saipb.RemoveL2McGroupMemberRequest{
+			Oid: p.oid,
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Update internal data
+	delete(mg.groups, req.GetOid())
 	return &saipb.RemoveL2McGroupResponse{}, nil
 }
 
-// TODO: Implement this.
-func (mg *l2mcGroup) RemoveL2McGroupMember(context.Context, *saipb.RemoveL2McGroupMemberRequest) (*saipb.RemoveL2McGroupMemberResponse, error) {
+func (mg *l2mcGroup) RemoveL2McGroupMember(ctx context.Context, req *saipb.RemoveL2McGroupMemberRequest) (*saipb.RemoveL2McGroupMemberResponse, error) {
+	// Remove table entry.
+	r := fwdconfig.TableEntryRemoveRequest(mg.dataplane.ID(), L2MCGroupTable).AppendEntry(
```

Comment:
shouldn't you be modifying the entry instead, removing a single member from the actions

Reply:
Refactored the code.